python/ray/tune/examples/xgboost_dynamic_resources_example.py

```diff
@@ -0,0 +1,228 @@
+        _, _, raw_model = pickle.load(inputFile)
+    best_bst.load_model(bytearray(raw_model))
+    accuracy = 1. - analysis.best_result["eval-logloss"]
+    print(f"Best model parameters: {analysis.best_config}")
+    print(f"Best model total accuracy: {accuracy:.4f}")
+    return best_bst
+
+
+def tune_xgboost():
+    search_space = {
+        # You can mix constants with search space objects.
+        "objective": "binary:logistic",
+        "eval_metric": ["logloss", "error"],
+        "max_depth": 9,
+        "learning_rate": 1,
+        "min_child_weight": tune.grid_search([2, 3]),
+        "subsample": tune.grid_search([0.8, 0.9]),
+        "colsample_bynode": tune.grid_search([0.8, 0.9]),
+        "random_state": 1,
+        "num_parallel_tree": 2000,
+    }
+    # This will enable aggressive early stopping of bad trials.
+    base_scheduler = ASHAScheduler(
+        max_t=16,  # 16 training iterations
+        grace_period=1,
+        reduction_factor=2)
+
+    def example_resources_allocation_function(
+            trial_runner: "trial_runner.TrialRunner", trial: Trial,
+            result: Dict[str, Any],
+            base_trial_resource: Union[PlacementGroupFactory, Resources]
+    ) -> Union[None, PlacementGroupFactory, Resources]:
+        """This is a basic example of a resource allocating function.
+
+        The function naively balances available CPUs over live trials.
+
+        This function returns a new ``PlacementGroupFactory`` with updated
+        resource requirements, or None. If the returned
+        ``PlacementGroupFactory`` is equal by value to the one the
+        trial has currently, the scheduler will skip the update process
+        internally (same with None).
+
+        See :func:`evenly_distribute_cpus_gpus` for a more complex,
+        robust approach.
+
+        Args:
+            trial_runner (TrialRunner): Trial runner for this Tune run.
+                Can be used to obtain information about other trials.
+            trial (Trial): The trial to allocate new resources to.
+            result (Dict[str, Any]): The latest results of trial.
+            base_trial_resource (Union[PlacementGroupFactory, Resources]):
+                Base trial resources as defined in
+                ``tune.run(resources_per_trial)``
+        """
+
+        # Don't bother if this is just the first iteration
+        if result["training_iteration"] < 1:
+            return None
+
+        # Assume that the number of CPUs cannot go below what was
+        # specified in tune.run
+        min_cpu = base_trial_resource.required_resources.get("CPU", 0)
+
+        # Get the number of CPUs available in total (not just free)
+        total_available_cpus = (
+            trial_runner.trial_executor._avail_resources.cpu)
+
+        # Divide the free CPUs among all live trials
+        cpu_to_use = max(
+            min_cpu,
+            total_available_cpus // len(trial_runner.get_live_trials()))
```

Comment:
Just to be sure: Is the current trial guaranteed to be in `get_live_trials()` (i.e., is `len(...) > 0` guaranteed)?

Reply:
Also, should we enforce 1 as a global minimum for numbers of CPUs?

Reply:
> Just to be sure: Is the current trial guaranteed to be in get_live_trials() (i.e., is len(...) > 0 guaranteed)?
Yes, it's not possible for a terminated or errored trial to be here

> Also, should we enforce 1 as a global minimum for numbers of CPUs?
I think it's fine to just leave it as is (the minimum being defined by base trial resources), though I suppose if they are not given then we should default to 1 CPU as a minimum